lua/core/menu.lua

```diff
@@ -1145,7 +1146,119 @@ m.deinit[pRESET] = function() end
 
 
 -----------------------------------------
+-- UPDATE
+m.update = {}
+m.update.url = ''
+m.update.version = ''
+
+local function check_newest()
+  print("checking for update")
+  m.update.url = util.os_capture( [[curl -s \
+      https://api.github.com/repos/monome/norns/releases/latest \
+      | grep "browser_download_url.*" \
+      | cut -d : -f 2,3 \
+      | tr -d \"]])
+  print(m.update.url)
+  m.update.version = m.update.url:match("(%d%d%d%d%d%d)")
+  print("available version "..m.update.version)
+end
+
+local function get_update()
+  m.update.message = "preparing..."
+  menu.redraw()
+  pcall(cleanup) -- shut down script
+  norns.script.clear()
+  print("shutting down audio...")
+  os.execute("sudo systemctl stop norns-jack.service") -- disable audio
+  print("clearing old updates...")
+  os.execute("sudo rm -rf /home/we/update/*") -- clear old updates
+  m.update.message = "downloading..."
+  menu.redraw()
+  print("starting download...")
+  os.execute("wget -T 180 -q -P /home/we/update/ " .. m.update.url) --download
+  m.update.message = "unpacking update..."
+  menu.redraw()
+  print("checksum validation...")
+  m.update.message = "checksum validation..."
+  local checksum = util.os_capture("cd /home/we/update; sha256sum -c /home/we/update/*.sha256 | grep OK")
+  if checksum:match("OK") then
+    print("unpacking...")
+    os.execute("tar xzvf /home/we/update/*.tgz -C /home/we/update/")
+    m.update.message = "running update..."
+    menu.redraw()
+    print("running update...")
+    os.execute("/home/we/update/"..m.update.version.."/update.sh")
+    m.update.message = "complete."
+    menu.redraw()
+    print("update complete.")
+  else
+    print("update failed.")
+    m.update.message = "update failed."
+    menu.redraw()
+  end
+end
 
+m.key[pUPDATE] = function(n,z)
+  if m.update.stage=="init" and z==1 then
+    menu.set_page(pSYSTEM)
+    menu.redraw()
+  elseif m.update.stage=="confirm" then
+    if n==2 and z==1 then
+      menu.set_page(pSYSTEM)
+      menu.redraw()
+    elseif n==3 and z==1 then
+      m.update.stage="update"
+      get_update()
+      m.update.stage="done"
+    end
+  elseif m.update.stage=="done" and z==1 then
+    print("shutting down.")
+    m.update.message = "shutting down."
+    menu.redraw()
+    os.execute("sleep 0.5; sudo shutdown now")
+  end
+end
+
+
+m.enc[pUPDATE] = function(n,delta) end
+
+m.redraw[pUPDATE] = function()
+  screen.clear()
+  screen.level(15)
+  screen.move(64,40)
+  if m.update.stage == "init" then
+    screen.text_center(m.update.message)
+  elseif m.update.stage == "confirm" then
+    screen.text_center("update found: "..m.update.version)
+    screen.move(64,50)
+    screen.text_center("install?")
+  elseif m.update.stage == "update" then
+    screen.text_center(m.update.message)
+  end
+  screen.update()
+end
+
+m.init[pUPDATE] = function()
+  m.update.stage = "init"
+
+  local ping = util.os_capture("ping -c 1 github.com | grep failure")
```

Comment:
it occurs to me that yet another option would be to check `ping` exit codes.

no lua handy but i *think* something like this might be equivalent?

```lua
var code = os.execute('ping -c 1 github.com')
if code == 0 then check_newest() end
```

(success is 0 and failure is anything else.)

Reply:
yeah, that works.

```
os.execute('ping -c 1 github.com')
PING github.com (192.30.253.112) 56(84) bytes of data.
64 bytes from lb-192-30-253-112-iad.github.com (192.30.253.112): icmp_seq=1 ttl=49 time=25.4 ms
--- github.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 25.470/25.470/25.470/0.000 ms
true	exit	0

os.execute('ping -c 1 githubssssss.com')
ping: githubssssss.com: Name or service not known
nil	exit	2
```

seems like it returns tables

Reply:
interesting and not what i expected!